**Assignment 9.3 - Calculate Probability of a Model Ensemble**

In [6]:
##View os and imports
import pandas as pd
import numpy as np
import scipy
from scipy.stats import binom
import os
print (os)

<module 'os' from 'C:\\Users\\14029\\anaconda3\\lib\\os.py'>


In this analysis I will take the probability for three model ensembles to assisnt in calculating the mean and variance using the binom function in scipy. 

In [14]:
##Calc probability of binomial assignment 1
probability = binom.pmf(k = 5, n = 11, p = 0.2) ##K = number of models / 2

print("Probability =", probability)

Probability = 0.038755368960000015


In [17]:
##Call for n and p 
n = 11
p = 0.2

##create list for r, n and p values
r_values = list(range(n+11))

##using binom.stats for mean and variance 
mean, var = binom.stats(n, p)

dist = [binom.pmf(r, n, p) for r in r_values]

print("r\tp(r)")
for i in range(n+1):
    print(str(r_values[i]) + "\t" + str(dist[i]))
      
##Print for mean and variance
print("Mean = " + str(mean))
print("Variance = " + str(var))

r	p(r)
0	0.08589934592000005
1	0.23622320128000018
2	0.2952790016
3	0.22145925120000007
4	0.11072962559999995
5	0.038755368960000015
6	0.009688842240000004
7	0.0017301504000000016
8	0.0002162688000000003
9	1.8022400000000016e-05
10	9.011200000000006e-07
11	2.0480000000000012e-08
Mean = 2.2
Variance = 1.7600000000000002


The probability for assignment 1 is shown as a variance of 1.76

In [18]:
##Calc probability of binomial assignment 2
probability = binom.pmf(k = 5, n = 11, p = 0.49) ##K = number of models / 2

print("Probability =", probability)

Probability = 0.22963782894651696


In [19]:
##Call for n and p 
n = 11
p = 0.49

##create list for r, n and p values
r_values = list(range(n+11))

##using binom.stats for mean and variance 
mean, var = binom.stats(n, p)

dist = [binom.pmf(r, n, p) for r in r_values]

print("r\tp(r)")
for i in range(n+1):
    print(str(r_values[i]) + "\t" + str(dist[i]))
      
##Print for mean and variance
print("Mean = " + str(mean))
print("Variance = " + str(var))

r	p(r)
0	0.0006071163615208264
1	0.0064163866443083415
2	0.030823818193245995
3	0.08884512302759136
4	0.17072200111184202
5	0.22963782894651696
6	0.22063242388979068
7	0.15141440855181712
8	0.07273829430430429
9	0.02329527072490786
10	0.004476346139296035
11	0.00039098210485829875
Mean = 5.39
Variance = 2.7489


The probability for assignment 2 is shown as a variance of 2.75

In [22]:
##Calc probability of binomial assignment 3
probability = binom.pmf(k = 10, n = 21, p = 0.49) ##K = number of models / 2

print("Probability =", probability)

Probability = 0.1708668834234244


In [23]:
##Call for n and p 
n = 21
p = 0.49

##create list for r, n and p values
r_values = list(range(n+11))

##using binom.stats for mean and variance 
mean, var = binom.stats(n, p)

dist = [binom.pmf(r, n, p) for r in r_values]

print("r\tp(r)")
for i in range(n+1):
    print(str(r_values[i]) + "\t" + str(dist[i]))
      
##Print for mean and variance
print("Mean = " + str(mean))
print("Variance = " + str(var))

r	p(r)
0	7.227260322084054e-07
1	1.4582060532204868e-05
2	0.00014010215021138044
3	0.0008525170055346097
4	0.0036858823474584536
5	0.012040549001697619
6	0.030848988291931202
7	0.0635126229539758
8	0.10678838075104752
9	0.14820086827541915
10	0.1708668834234244
11	0.1641662213283881
12	0.13144027524658533
13	0.08742859937216295
14	0.04800001534157972
15	0.021521575506093935
16	0.007754097057342665
17	0.0021911808293528837
18	0.00046783381541739205
19	7.09716933605329e-05
20	6.8188489699335564e-06
21	3.1197348228454223e-07
Mean = 10.29
Variance = 5.2479


The probability for assignment 2 is shown as a variance of 5.25

The mean can be calculated and verified by simply taking the the number of models and dividing it by the p value for a quick means of double checking and ensuring the formula is working as desired. 

For example assignment 3 has 21 models and a p value of .49. 49% of 21 is 10.29. 